lib/srv/reexec.go

```diff
@@ -740,63 +578,296 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
+		}
+	}
+
+	// Ensure that the working directory is one that the local user has access to.
+	if err := os.Chdir(homeDir); err != nil {
+		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to set working directory for networking process")
+	}
+
+	// Build request listener from first extra file that was passed to command.
 	ffd := os.NewFile(FirstExtraFile, "listener")
 	if ffd == nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.BadParameter("missing socket fd")
 	}
 
-	conn, err := uds.FromFile(ffd)
-	ffd.Close()
+	parentConn, err := uds.FromFile(ffd)
+	_ = ffd.Close()
 	if err != nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	writeErrorToConn := func(conn io.Writer, err error) {
+		conn.Write([]byte(err.Error()))
+	}
+
+	// Maintain a list of file paths to cleanup at the end of the process. This
+	// ensures that file cleanup is handled by the child in cases where the parent
+	// fails to cleanup due to filesystem namespace discrepancy (pam_namespace)
+	var filePathsToCleanup []string
+	defer func() {
+		for _, path := range filePathsToCleanup {
+			os.Remove(path)
+		}
+	}()
+
+	// parentConn is a datagram Unix socket, which is not connection oriented
+	// and thus won't unblock when the parent closes its side of the connection.
+	// Instead we use an interrupt signal from the parent process to unblock.
+	go func() {
+		_, _ = terminatefd.Read(make([]byte, 1))
+		parentConn.Close()
+	}()
+
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
+				// parent connection closed, process should exit.
 				return errorWriter, teleport.RemoteCommandSuccess, nil
 			}
-			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
+			writeErrorToConn(parentConn, trace.Wrap(err, "error reading networking request from parent"))
+			continue
 		}
-		addr := string(buf[:n])
+
 		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
+			writeErrorToConn(parentConn, trace.BadParameter("networking request requires a control file"))
 			continue
 		}
 
-		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
+		requestConn, err := uds.FromFile(fbuf[0])
+		_ = fbuf[0].Close()
+		if err != nil {
+			writeErrorToConn(parentConn, trace.Wrap(err, "failed to get a connection from control file"))
+			continue
+		}
+
+		var req networking.Request
+		if err := json.Unmarshal(buf[:n], &req); err != nil {
+			writeErrorToConn(parentConn, trace.Wrap(err, "error parsing networking request"))
```

Comment:
This can be an error returned by the request I think?
```suggestion
			writeErrorToConn(requestConn, trace.Wrap(err, "error parsing networking request"))
```

Reply:
But the requestConn won't respect the error message boundaries since it's a stream, right?

Reply:
That's true - but we can just read the whole connection until the EOF after we try to read the file descriptor; if we don't do that we're kinda just relying on luck to get decent error messages for the user (the message is propagated into the response to the direct-tcpip request, right?)